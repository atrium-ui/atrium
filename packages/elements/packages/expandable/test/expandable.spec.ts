import { test, expect } from "bun:test";

const NODE_NAME = "a-expandable";

test("import element", async () => {
  const { Expandable } = await import("@sv/elements/expandable");
  expect(Expandable).toBeDefined();

  // is defined in custom element registry
  expect(customElements.get(NODE_NAME)).toBeDefined();

  // is constructable
  expect(new Expandable()).toBeInstanceOf(Expandable);

  const html = `<${NODE_NAME} />`;
  const ele = document.createElement("div");
  ele.innerHTML = html;

  expect(ele.children[0]).toBeInstanceOf(Expandable);
});

test("change event", async () => {
  const ele = await newExpandable();

  let changed = false;
  ele.addEventListener("change", () => {
    changed = true;
  });

  ele.opened = true;
  expect(changed).toBeFalse();

  // only from use interaction
  ele.open();
  expect(changed).toBeTrue();
});

test("open and close", async () => {
  const ele = await newExpandable();

  expect(ele.opened).toBe(false);

  expect(ele.outerHTML).toMatchSnapshot("closed");

  ele.open();
  expect(ele.opened).toBe(true);

  ele.close();
  expect(ele.opened).toBe(false);

  open(ele);
  expect(ele.opened).toBe(true);

  expect(ele.outerHTML).toMatchSnapshot("opened");
});

test("initialy opened", async () => {
  const ele = await newExpandableOpened();

  expect(ele.opened).toBe(true);
});

test("unique ids", async () => {
  const ele = await newExpandableOpened();
  const ele2 = await newExpandableOpened();

  expect(ele._id_toggle).not.toEqual(ele2._id_toggle);
  expect(ele._id_content).not.toEqual(ele2._id_content);
});

test("aria attributes", async () => {
  const ele = await newExpandable();

  const trigger = ele.trigger;

  expect(trigger.getAttribute("aria-controls")).toBe(ele._id_content);
  expect(trigger.id).toBe(ele._id_toggle);

  const content = ele.content;
  expect(content?.getAttribute("aria-labelledby")).toBe(ele._id_toggle);

  open(ele);

  expect(trigger.getAttribute("aria-expanded")).toBe("true");
  expect(content.getAttribute("aria-hidden")).toBe("false");
});

test("no change event when attribute is changed", async () => {
  const ele = await newExpandable();

  expect(ele.opened).toBe(false);

  let changeEvent = false;
  ele.addEventListener("change", () => {
    changeEvent = true;
  });

  ele.setAttribute("opened", "true");
  expect(ele.opened).toBe(true);

  await sleep(1);

  expect(changeEvent).toBe(false);
});

async function newExpandable() {
  await import("@sv/elements/expandable");
  const ele = document.createElement("div");

  ele.innerHTML = `
    <a-expandable>
      <button slot="toggle" type="button">
        <div class="headline">Title</div>
      </button>

      <div>Content</div>
    </a-expandable>
  `;

  const expandable = ele.querySelector("a-expandable");
  document.body.appendChild(ele);
  const ex = expandable as any;
  ex.onSlotChange();
  return ex;
}

async function newExpandableOpened() {
  await import("@sv/elements/expandable");
  const ele = document.createElement("div");

  ele.innerHTML = `
    <a-expandable opened>
      <button slot="toggle" type="button">
        <div class="headline">Title</div>
      </button>

      <div>Content</div>
    </a-expandable>
  `;

  const expandable = ele.querySelector("a-expandable");
  document.body.appendChild(ele);
  const ex = expandable as any;
  ex.onSlotChange();
  return ex;
}

function open(ele) {
  ele.onClick({ target: ele.trigger });
}

function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
