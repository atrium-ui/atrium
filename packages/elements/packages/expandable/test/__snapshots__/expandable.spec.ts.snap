// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`open and close: closed 1`] = `
"<a-expandable direction="down">
      <button slot="toggle" type="button" aria-controls="expandable_content_4" id="expandable_toggle_4">
        <div class="headline">Title</div>
      </button>

      <div role="region" id="expandable_content_4" aria-labelledby="expandable_toggle_4">Content</div>
    </a-expandable>"
`;

exports[`open and close: opened 1`] = `
"<a-expandable direction="down">
      <button slot="toggle" type="button" aria-controls="expandable_content_4" id="expandable_toggle_4" aria-expanded="true">
        <div class="headline">Title</div>
      </button>

      <div role="region" id="expandable_content_4" aria-labelledby="expandable_toggle_4" aria-hidden="false">Content</div>
    </a-expandable>"
`;

exports[`open and close: closed 2`] = `
"<a-expandable direction="down">
      <button slot="toggle" type="button" aria-controls="expandable_content_13" id="expandable_toggle_13">
        <div class="headline">Title</div>
      </button>

      <div role="region" id="expandable_content_13" aria-labelledby="expandable_toggle_13">Content</div>
    </a-expandable>"
`;

exports[`open and close: opened 2`] = `
"<a-expandable direction="down">
      <button slot="toggle" type="button" aria-controls="expandable_content_13" id="expandable_toggle_13" aria-expanded="true">
        <div class="headline">Title</div>
      </button>

      <div role="region" id="expandable_content_13" aria-labelledby="expandable_toggle_13" aria-hidden="false">Content</div>
    </a-expandable>"
`;

exports[`open and close: closed 3`] = `
"<a-expandable direction="down">
      <button slot="toggle" type="button" aria-controls="expandable_content_22" id="expandable_toggle_22">
        <div class="headline">Title</div>
      </button>

      <div role="region" id="expandable_content_22" aria-labelledby="expandable_toggle_22">Content</div>
    </a-expandable>"
`;

exports[`open and close: opened 3`] = `
"<a-expandable direction="down">
      <button slot="toggle" type="button" aria-controls="expandable_content_22" id="expandable_toggle_22" aria-expanded="true">
        <div class="headline">Title</div>
      </button>

      <div role="region" id="expandable_content_22" aria-labelledby="expandable_toggle_22" aria-hidden="false">Content</div>
    </a-expandable>"
`;

exports[`open and close: closed 4`] = `
"<a-expandable direction="down">
      <button slot="toggle" type="button" aria-controls="expandable_content_31" id="expandable_toggle_31">
        <div class="headline">Title</div>
      </button>

      <div role="region" id="expandable_content_31" aria-labelledby="expandable_toggle_31">Content</div>
    </a-expandable>"
`;

exports[`open and close: opened 4`] = `
"<a-expandable direction="down">
      <button slot="toggle" type="button" aria-controls="expandable_content_31" id="expandable_toggle_31" aria-expanded="true">
        <div class="headline">Title</div>
      </button>

      <div role="region" id="expandable_content_31" aria-labelledby="expandable_toggle_31" aria-hidden="false">Content</div>
    </a-expandable>"
`;
